06_gpu_and_ml/sam/sam_example.py

```diff
@@ -0,0 +1,178 @@
+# ---
+# cmd: ["modal", "run", "06_gpu_and_ml.sam.app"]
+# deploy: true
+# ---
+
+import modal
+
+from .helper import show_anns, show_mask, show_points
+
+# # Run Facebook's Segment Anything Model (SAM) on Modal
+
+# This example demonstrates how to deploy Facebook's Segment Anything Model (SAM)
+# on Modal. SAM is a powerful, flexible image segmentation model that can be used
+# for various computer vision tasks like object detection, instance segmentation,
+# and even as a foundation for more complex computer vision applications.
+# For more information, see: https://github.com/facebookresearch/segment-anything
+
+# # Setup
+
+# First, we set up the Modal image with the necessary dependencies, including PyTorch,
+# OpenCV, and Torchvision. We also install the Segment Anything Model (SAM) library.
+
+MODEL_TYPE = "vit_h"
+
+image = (
+    modal.Image.debian_slim(python_version="3.10")
+    .apt_install("git", "wget", "libgl1-mesa-glx")
+    .pip_install(
+        "torch",
+        "torchvision",
+        "opencv-python",
+        "pycocotools",
+        "matplotlib",
+        "onnxruntime",
+        "onnx",
+        "git+https://github.com/facebookresearch/segment-anything.git",
+    )
+)
+
+app = modal.App("sam-app", image=image)
+
+# # Model definition
+
+# Next, we define the Model class that will handle SAM operations
+
+
+# We use @modal.build() and @modal.enter() decorators here for optimization:
+# @modal.build() ensures this method runs during the container build process,
+# downloading the model only once and caching it in the container image.
+# @modal.enter() makes sure the method runs only once when a new container starts,
+# initializing the model and moving it to GPU.
+# The upshot is that model downloading and initialization only happen once upon container startup.
+# This significantly reduces cold start times and improves performance.
+@app.cls(gpu="any", timeout=600)
+class Model:
+    def __init__(self, model_type=MODEL_TYPE):
+        self.model_type = model_type
+
+    # # Model initialization
+    @modal.build()
+    @modal.enter()
+    def download_model_to_folder(self):
+        import subprocess
+
+        from segment_anything import (
+            SamAutomaticMaskGenerator,
+            SamPredictor,
+            sam_model_registry,
+        )
+
+        # SAM offers different model sizes (ViT-H, ViT-L, ViT-B)
```

Comment:
Can we use SAM2 instead of 1? https://github.com/facebookresearch/sam2/blob/05d9e57fb3945b10c861046c1e6749e2bfc258e3/README.md?plain=1#L62